### DISCUSSION TITLE
Support `undefined` as a default value in Parsers

### DISCUSSION THREAD

**Discussion by taylorjohnwood** [47ng/nuqs#595] [Ideas]
For better or worse, some applications have a strict semantic difference between `null` and `undefined`. Would it be possible to support `undefined` as a default for parsers?

e.g
```
parseAsInteger.withDefault(undefined) // Allow this, currently due to typing it does not work.
```

I'm happy to raise a PR for this if people agree it belongs in Nuqs!

**Comment by franky47** [comment]
You could wrap your useQueryState hook to get this behaviour:

```ts
function useUndefinedQueryState() {
  const [state, setState] = useQueryState('my-key')
  return [state ?? undefined, setState] as const
}
```

You would still have to set the value to `null` to remove it from the URL though, which might cause confusion as you'd get `undefined` back.

If you want to try your hand at a PR, the hardest part will be updating the type definitions, and the `clearOnDefault` option runtime code might require some tweaking to properly handle `undefined`.

**Reply by taylorjohnwood** [reply]
Yes good suggestion :)  I've currently wrapped a lot of my parsing to get this behavior, but as you mentioned removing query params requires `null`. My other complications with wrapping are:

1. We reuse the same parsers across multiple pages and there is a chance that someone one accidentally forgets to wrap them.
2. I'm usually parsing objects so the wrapping is a little more tedious
```
const [coordinates, setCoordinates] = useQueryStates(
  {
    lat: parseAsFloat.withDefault(45.18),
    lng: parseAsFloat.withDefault(5.72),
    ... // many more params
  },
)

const coordinatesWithUndefined = {
  lat: lat ?? undefined,
  lng: lng ?? undefined,
  ... // repeat for all query params
} 
```

I'm going to raise a PR soon I think. Thanks for your quick reply

**Reply by franky47** [reply]
Sounds good, please review the [contribution guidelines](https://github.com/47ng/nuqs/blob/next/CONTRIBUTING.md) to get started. Adding typing tests (*.test-d.ts) will be key here to make sure type updates behave correctly.
